src/__tests__/integration/basic/constructor.test.ts

```diff
@@ -0,0 +1,225 @@
+
+                expect(state.calls).toBeDefined();
+                expect(state.calls).toEqual(['__init', 'nop']);
+
+                await contract.writeInteraction({ function: 'nop' });
+                await contract.writeInteraction({ function: 'nop' });
+
+                const { cachedValue: { state: state2 } } = await contract.readState();
+
+                expect(state2.calls).toBeDefined();
+                expect(state2.calls).toEqual(['__init', 'nop', 'nop', 'nop']);
+            })
+        });
+
+        describe('Constructor has access to all smartweave globals', () => {
+            it('should assign as caller deployer of contract', async () => {
+                const contract = await deployContract({});
+
+                const { cachedValue: { state } } = await contract.readState();
+
+                expect(state.caller).toEqual(walletAddress);
+                expect(state.caller2).toEqual(walletAddress);
+
+                await contract.writeInteraction({ function: 'nop' });
+                const { cachedValue: { state: state2 } } = await contract.readState();
+                expect(state2.caller).toEqual(walletAddress);
+                expect(state2.caller2).toEqual(walletAddress);
+            });
+
+            it('should work with KV', async () => {
+                const contract = await deployContract({});
+
+                await contract.readState();
+                const { cachedValue: kv } = await contract.getStorageValues(
+                    ['__init']
+                );
+                expect(kv.get('__init')).toEqual(contract.txId());
+            });
+        });
+
+        it('should rollback KV and state', async () => {
+            const contract = await deployContract({ addToState: { fail: true } });
+
+            await expect(contract.readState()).rejects.toThrowError();
+            const { cachedValue: kv } = await contract.getStorageValues(
+                ['__init']
+            );
+            expect(kv.get('__init')).toEqual(undefined);
+        });
+
+        it('should fail to call __init function explicit', async () => {
+            const contract = await deployContract({});
+
+            await expect(contract.writeInteraction({ function: '__init' }, { strict: true })).rejects.toThrowError();
+        });
+
+        it('should properly apply modifications from __init', async () => {
+            const contract = await deployContract({});
+
+            const { cachedValue: { state } } = await contract.readState();
+
+            expect(state.counter).toBe(2);
+
+            await contract.writeInteraction({ function: 'nop' })
+            await contract.writeInteraction({ function: 'nop' })
+
+            const { cachedValue: { state: state2 } } = await contract.readState();
+            expect(state2.counter).toStrictEqual(2);
+        });
+
+        describe.skip('TODO: doesn"t work internal writes', () => {
```

Comment:
Here is my biggest nightmare, internal writes doesn't work.
I'm not sure why, maybe because `contractDefinition.contractTx` doesn't have assigned `sortKey` and internalWrites doesn't know if it should evaluate it

Reply:
Internasl writes won't be working in constructor